File: Práctica4/lib/NumericDamage.rb
#encoding:utf-8

require_relative 'Damage'
require_relative 'NumericDamageToUI'

module Deepspace
  # Clase que representa un daño numérico
  class NumericDamage < Damage
    public_class_method :new

    # Constructor
    # @param _nWeapons [Integer] numero de armas que se pierden
    # @param _nShields [Integer] numero de escudos que se pierden
    def initialize(_nWeapons, _nShields)
      super(_nShields)

      # @!attribute [Integer] numero de armas que se pierden
      @nWeapons = _nWeapons
    end

    # Getters
    attr_reader :nWeapons

    # Comprube si el daño hace efecto
    # @return [Boolean] true en caso afirmativo, false en caso contrario
    # --Overriden
    def hasNoEffect
      (super && @nWeapons == 0)
    end

    # Crea un parametro ajustado a las instancias pasadas como argumentos
    # @param w [Array<Weapon>] el array de armas a ajustar
    # @param s [Array<ShieldBooster>] el array de escudos a ajustar
    # @return [NumericDamage] la copia ajustada
    def adjust(w, s)
      self.class.new([@nWeapons, w.length].min, super(s))
    end

    # Devuelve una copia de la instancia actual
    # @return [NumericDamage] copia
    def copy
      self.class.new(nWeapons, nShields)
    end

    # Elimina un arma
    # Decrementa el contador de armas en uno
    # @param w [Weapon] arma
    def discardWeapon(w)
      if @nWeapons > 0
        @nWeapons -= 1
      end
    end

    # String representation of the object
    # @return [String] string representation
    def to_s
      message = "[Numeric Damage] -> Weapons: #{@nWeapons}, Shields: #{@nShields}"
      message
    end

    # To UI
    def getUIversion
      NumericDamageToUI.new(self)
    end
  end
end

File: Práctica5/java/src/deepspace/NumericDamage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 * NumericDamage está compuesto de:
 * - número de escudos a perder
 * - número de armas a perder
 */
public class NumericDamage extends Damage {
    /**
     * Número de armas a perder
     */
    private int nWeapons;
    
    /**
     * Constructor
     * @param _nWeapons número de armas a perder
     * @param _nShields número de escudos a perder
     */
    public NumericDamage(int _nWeapons, int _nShields) {
        super(_nShields);
        nWeapons = _nWeapons;
    }
    
    // Getters
    
    /**
     * Copy getter
     * @return copia de la instancia
     */
    @Override
    public NumericDamage copy(){
        return new NumericDamage(nWeapons, getNShields());
    }
    
    /**
     * Getter de nWeapons
     * @return nWeapons
     */
    public int getNWeapons(){
        return nWeapons;
    }
    
    /**
     * Comprueba si el daño afecta o no
     * @return true en caso de no afectar, false en caso contrario
     */
    @Override
    public boolean hasNoEffect(){
        return getNShields() + nWeapons == 0;
    }
    
    // Setters
    
    /**
     * Crea un objeto ajustado 
     * @param w armas a ajustar
     * @param s escudos a ajustar
     * @return copia del objecto ajustado
     */
    @Override
    public NumericDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s){
        int shields = Integer.min(s.size(), getNShields());
        return new NumericDamage(Integer.min(nWeapons, w.size()), shields);
    }
    
    /**
     * Elimina un arma, decrementando el contador en uno
     * @param w el arma a eliminar
     */
    @Override
    public void discardWeapon(Weapon w){
        if (nWeapons > 0)
            nWeapons--;
    }
    
    // -------------------------------------------------------------------------
    // String representation, UI version
    // -------------------------------------------------------------------------
    
    /**
     * String representation of the object.
     * @return string representation
     */
    @Override
    public String toString() {
        String message = "[NumericDamage] -> "
                + "Number of shields: " + getNShields()
                + ", Number of weapons: " + nWeapons;
        return message;
    }
    
    /**
     * To UI.
     */
    @Override
    public NumericDamageToUI getUIversion() {
        return new NumericDamageToUI(this);
    }
    
    
}
